neuralop/utils.py

```diff
@@ -1,42 +1,80 @@
+from typing import Union, Literal, Optional, List, Tuple, Generator
 import torch
+from torch import Tensor
 import wandb
 
+
 # normalization, pointwise gaussian
 class UnitGaussianNormalizer:
-    def __init__(self, x, eps=0.00001, reduce_dim=[0], verbose=True):
+    def __init__(
+        self,
+        x: Union[Tensor, Generator],
+        eps: float = 1e-6,
+        reduce_dim: List[int] = [0],
+        verbose: bool = True,
+    ):
         super().__init__()
-        n_samples, *shape = x.shape
-        self.sample_shape = shape
         self.verbose = verbose
         self.reduce_dim = reduce_dim
 
         # x could be in shape of ntrain*n or ntrain*T*n or ntrain*n*T
-        self.mean = torch.mean(x, reduce_dim, keepdim=True).squeeze(0)
-        self.std = torch.std(x, reduce_dim, keepdim=True).squeeze(0)
+        count, mean, std = self.compute_statistics(x, reduce_dim)
         self.eps = eps
-        
+        self.mean = mean
+        self.std = std
+        self.count = count
+
         if verbose:
-            print(f'UnitGaussianNormalizer init on {n_samples}, reducing over {reduce_dim}, samples of shape {shape}.')
-            print(f'   Mean and std of shape {self.mean.shape}, eps={eps}')
+            print(
+                f"UnitGaussianNormalizer init on {count}, reducing over {reduce_dim}."
+            )
+            print(f"   Mean and std of shape {self.mean.shape}, eps={eps}")
+
+    def compute_statistics(
+        self, x: Union[Tensor, Generator], reduce_dim: List[int] = [0]
+    ) -> Tuple[int, Tensor, Tensor]:
+        """
+        Compute mean and standard deviaion
+        """
+        if isinstance(x, Tensor):
```

Comment:
might be more robust to use torch.is_tensor(x)
Also we could just always use the second version?

Reply:
is_tensor is a wrapper for isinstance(x, tensor).